### PR TITLE
8297264: C2: Cast node is not processed again in CCP and keeps a wrong too narrow type which is later replaced by top

### PR DESCRIPTION
**Pull request by chhagedorn**
image

During CCP, we optimize the type of `348 CastII` in `CastIINode::Value()`: It matches the `CmpI/If` pattern because the current type of `119 Phi` is a constant int:
https://github.com/openjdk/jdk/blob/9f24a6f43c6a5e1fa92275e0a87af4f1f0603ba3/src/hotspot/share/opto/castnode.cpp#L213-L215

Later in CCP, the type of `119 Phi` is updated and is no longer a constant but `348 CastII` is not processed anymore during CCP and keeps its wrong too narrow type. We apply more loop opts and at some point, the `CastII` is replaced by top because the input type is outside of the wrong type range of the `CastII`. Some data nodes are folded and the graph is left in a broken state and we assert during GCM.

I propose to add a `CastII` node back to the CCP worklist if we find such a `Cmp/If` pattern to ensure that the `CastII` type is correctly set during CCP.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297264](https://bugs.openjdk.org/browse/JDK-8297264): C2: Cast node is not processed again in CCP and keeps a wrong too narrow type which is later replaced by top

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [f9dcf645](https://git.openjdk.org/jdk/pull/11448/files/f9dcf6454664afb503068caf5e5893c5503ec8f3)
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [f9dcf645](https://git.openjdk.org/jdk/pull/11448/files/f9dcf6454664afb503068caf5e5893c5503ec8f3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11448/head:pull/11448` \
`$ git checkout pull/11448`

Update a local copy of the PR: \
`$ git checkout pull/11448` \
`$ git pull https://git.openjdk.org/jdk pull/11448/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11448`

View PR using the GUI difftool: \
`$ git pr show -t 11448`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11448.diff">https://git.openjdk.org/jdk/pull/11448.diff</a>

</details>

### CODE REVIEW

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. FTR, the verification code I proposed in [JDK-8257197](https://bugs.openjdk.org/browse/JDK-8257197) would catch this.

**Comment by chhagedorn** [comment]
> Looks good to me. FTR, the verification code I proposed in [JDK-8257197](https://bugs.openjdk.org/browse/JDK-8257197) would catch this.

Thanks Tobias for your review! Yes, that's a good point. I think it would be good to get this verification code in at some point to catch such issues with CCP earlier.

File: src/hotspot/share/opto/phaseX.cpp [hunk 1]
```diff
@@ -1950,6 +1951,22 @@ void PhaseCCP::push_and(Unique_Node_List& worklist, const Node* parent, const No
...
+// of the CmpI (i.e. 'parent') is no longer constant. The type of the CastII must be widened in this case.
+void PhaseCCP::push_cast_ii(Unique_Node_List& worklist, const Node* parent, const Node* use) const {
+  if (use->Opcode() == Op_CmpI && use->in(2) == parent) {
+    Node* other_cmp_input = use->in(1);
+    for (DUIterator_Fast imax, i = other_cmp_input->fast_outs(imax); i < imax; i++) {
+      Node* cast_ii = other_cmp_input->fast_out(i);
+      if (cast_ii->is_CastII()) {
+        push_if_not_bottom_type(worklist, cast_ii);
+      }
+    }
```
Comment (robcasloz) [inline src/hotspot/share/opto/phaseX.cpp:1967]: Would it make sense to assume there is at most one `CastII` output and replace the loop with a call to the auxiliary function `Node* Node::find_out_with(int opcode)`?

File: test/hotspot/jtreg/compiler/c2/TestCastIIWrongTypeCCP.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8297264
+ * @summary Test that CastII nodes are added to the CCP worklist if they could or could have been
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/c2/TestCastIIWrongTypeCCP.java:27]:
```suggestion
 * @summary Test that CastII nodes are added to the CCP worklist if they could have been
```

File: test/hotspot/jtreg/compiler/c2/TestCastIIWrongTypeCCP.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
+/*
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/c2/TestCastIIWrongTypeCCP.java:1]: This file might fit better under `test/hotspot/jtreg/compiler/ccp`.

**Review by robcasloz — APPROVED** [review]
Looks good! Just a couple of questions/suggestions.

File: test/hotspot/jtreg/compiler/c2/TestCastIIWrongTypeCCP.java (hunk 3 above)
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/c2/TestCastIIWrongTypeCCP.java]: That's a good point, I'll move it over there.

File: src/hotspot/share/opto/phaseX.cpp (hunk 1 above)
Comment (chhagedorn) [reply src/hotspot/share/opto/phaseX.cpp:1967]: Unfortunately, in this test case, we have 2 `CastII` nodes that need to be re-added. So, I think that `find_out_with()` does not work here in general (even though in this test case, both `CastII` nodes would end up again on the worklist but that is probably not guaranteed in general).

**Review by chhagedorn — COMMENTED** [review]
Thanks Roberto for your review! I've updated the patch accordingly.

**Review by TobiHartmann — APPROVED** [review]
Still looks good.

File: src/hotspot/share/opto/phaseX.cpp (hunk 1 above)
Comment (robcasloz) [reply src/hotspot/share/opto/phaseX.cpp:1967]: I see, thanks for the explanation!

File: test/hotspot/jtreg/compiler/ccp/TestCastIIWrongTypeCCP.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8297264
+ * @summary Test that CastII nodes are added to the CCP worklist if they could have been
+ *          optimized due to a CmpI/If pattern.
+ * @run main/othervm -Xcomp -XX:CompileCommand=compileonly,compiler.c2.TestCastIIWrongTypeCCP::*
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/ccp/TestCastIIWrongTypeCCP.java:29]:
```suggestion
 * @run main/othervm -Xcomp -XX:CompileCommand=compileonly,compiler.ccp.TestCastIIWrongTypeCCP::*
```

File: test/hotspot/jtreg/compiler/ccp/TestCastIIWrongTypeCCP.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ *                   compiler.c2.TestCastIIWrongTypeCCP
+ */
+package compiler.c2;
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/ccp/TestCastIIWrongTypeCCP.java:32]:
```suggestion
package compiler.ccp;
```

File: test/hotspot/jtreg/compiler/ccp/TestCastIIWrongTypeCCP.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8297264
+ * @summary Test that CastII nodes are added to the CCP worklist if they could have been
+ *          optimized due to a CmpI/If pattern.
+ * @run main/othervm -Xcomp -XX:CompileCommand=compileonly,compiler.c2.TestCastIIWrongTypeCCP::*
+ *                   compiler.c2.TestCastIIWrongTypeCCP
```
Comment (robcasloz) [inline test/hotspot/jtreg/compiler/ccp/TestCastIIWrongTypeCCP.java:30]:
```suggestion
 *                   compiler.ccp.TestCastIIWrongTypeCCP
```

**Review by robcasloz — CHANGES_REQUESTED** [review]
Thanks for addressing my feedback, Christian! Just one thing, you probably want to update the test package name for consistency.

**Comment by chhagedorn** [comment]
Oh right, good catch, thanks! I've updated it with your suggestions.

**Review by robcasloz — APPROVED** [review]

**Pull request closed without merging**